src/jsonpatcherproxy.js

```diff
@@ -34,73 +39,89 @@ const JSONPatcherProxy = (function() {
+   * It allows to check get the tree metadata from within the tree. This allows to check if the object is proxified by this
+   * instance of JSONPatcherProxy, even when all you have is a reference that is someone else's proxy
+   * @param {JSONPatcherProxy} instance JSONPatcherProxy instance
+   * @param {Object} tree the affected object
+   * @param {String} key the effect property's name
+   */
+  function trapForGet(instance, tree, key) {
+    if (key === instance._metadataSymbol) {
+      return instance._treeMetadataMap.get(tree);
+    }
+    return Reflect.get(tree, key);
+  }
   /**
    * A callback to be used as the proxy set trap callback.
-   * It updates parenthood map if needed, proxifies nested newly-added objects, calls default callback with the changes occurred.
+   * It updates tree metadata map if needed, proxifies nested newly-added objects, calls default callback with the changes occurred.
    * @param {JSONPatcherProxy} instance JSONPatcherProxy instance
    * @param {Object} tree the affected object
    * @param {String} key the effect property's name
    * @param {Any} newValue the value being set
    */
   function trapForSet(instance, tree, key, newValue) {
     const pathToKey = getPathToTree(instance, tree) + '/' + escapePathComponent(key);
-    const subtreeMetadata = instance._treeMetadataMap.get(newValue);
-
-    if (instance._treeMetadataMap.has(newValue)) {
-      instance._parenthoodMap.set(subtreeMetadata.originalObject, { parent: tree, key });
-    }
-    /*
-        mark already proxified values as inherited.
-        rationale: proxy.arr.shift()
-        will emit
-        {op: replace, path: '/arr/1', value: arr_2}
-        {op: remove, path: '/arr/2'}
-
-        by default, the second operation would revoke the proxy, and this renders arr revoked.
-        That's why we need to remember the proxies that are inherited.
-      */
-    /*
-    Why do we need to check instance._isProxifyingTreeNow?
-
-    We need to make sure we mark revocables as inherited ONLY when we're observing,
-    because throughout the first proxification, a sub-object is proxified and then assigned to 
-    its parent object. This assignment of a pre-proxified object can fool us into thinking
-    that it's a proxified object moved around, while in fact it's the first assignment ever. 
-
-    Checking _isProxifyingTreeNow ensures this is not happening in the first proxification, 
-    but in fact is is a proxified object moved around the tree
-    */
-    if (subtreeMetadata && !instance._isProxifyingTreeNow) {
-      subtreeMetadata.inherited = true;
+    
+    if (isObject(newValue)) {
+      const subtreeMetadata = newValue[instance._metadataSymbol];
+      if (subtreeMetadata) {
+        if(subtreeMetadata.parent === tree && subtreeMetadata.keyInParent === key) {
+          //this is the same object that we already proxified, proxified now by someone else. In this case, remain silent
```

Comment:
"we" means this instance, not JSONPatcherProxy in general?

Is it necessary "proxified now by someone else" Isnt; it run also for `tree.key = tree.key`?

Reply:
>"we" means this instance, not JSONPatcherProxy in general?

Yes, "we" means "this instance of `JSONPatcherProxy`", because tree metadata is private per instance of `JSONPatcherProxy`. 

> Is it necessary "proxified now by someone else" Isnt; it run also for tree.key = tree.key?

You are right, it will also go here if it is simply reassigned at the same place.

I will clarify in the comment.